### PR TITLE
Mr/procure next

### PR DESCRIPTION
**Pull request by SeduroDotCom**

### CODE REVIEW

File: server/src/leihs/procurement/resources/viewers.clj [hunk 1]
```diff
@@ -1,38 +1,43 @@
...
-  (jdbc/execute! tx
-                 (-> (sql/delete-from [:procurement_category_viewers :pcv])
-                     (sql/merge-where [:= :pcv.category_id c-id])
-                     sql/format)))
+  (let [result (jdbc/execute-one! tx (-> (sql/delete-from :procurement_category_viewers :pcv)
+                                              (sql/where [:= :pcv.category_id [:cast c-id :uuid]])
+                                              sql-format))
+
+        res (spy (:update-count result))                    ;; TODO: fixme
+        result (spy (:next.jdbc/update-count result))]
```
Comment (SeduroDotCom) [inline server/src/leihs/procurement/resources/viewers.clj:28]: jdbc-next prefix needed

File: server/src/leihs/procurement/resources/template.clj [hunk 2]
```diff
@@ -14,64 +17,72 @@
...
-                     (sql/sset (validate-update-attributes tx tmpl))
-                     (sql/where [:= :procurement_templates.id (:id tmpl)])
-                     sql/format)))
+  (let [casted-tmpl (my-cast  tmpl)
+        casted-tmpl (validate-update-attributes tx casted-tmpl)]
+    (-> (jdbc/execute-one! tx (-> (sql/update :procurement_templates)
+                                  (sql/set  casted-tmpl)
+                                  (sql/where [:= :procurement_templates.id (:id casted-tmpl)])
+                                  sql-format))
+        :next.jdbc/update-count
```
Comment (SeduroDotCom) [inline server/src/leihs/procurement/resources/template.clj:56]: jdbc-next prefix needed

File: server/src/leihs/procurement/resources/request.clj [hunk 3]
```diff
@@ -497,21 +505,25 @@
...
-                                       sql/format))]
-        (= result '(1)))
+      #(let [result (jdbc/execute-one! tx (-> (sql/delete-from :procurement_requests)
+                                              (sql/where [:= :procurement_requests.id [:cast req-id :uuid]])
+                                              sql-format) {:builder-fn next.jdbc.result-set/as-unqualified-maps})
+             ;; TODO: FIXME, namespace shouldn't be needed here
+             result-count (spy (:update-count (spy result))) ;; fails
+             result-count (spy (:next.jdbc/update-count (spy result))) ;; works
+             ]
+         (= result-count 1))
```
Comment (SeduroDotCom) [inline server/src/leihs/procurement/resources/request.clj:515]: jdbc-next prefix needed

File: server/deps.edn [hunk 4]
```diff
@@ -13,6 +13,11 @@
         inflections/inflections {:mvn/version "0.13.2"}
         leihs-core/leihs-core {:local/root "./shared-clj"}
         org.clojure/tools.cli {:mvn/version "1.0.206"}
+
+        com.github.seancorfield/honeysql {:mvn/version "2.4.1066"}
+        com.github.seancorfield/next.jdbc {:mvn/version "1.3.894"}
+        logbug/logbug {:mvn/version "5.0.0"}
+
```
Comment (SeduroDotCom) [inline server/deps.edn:20]: TODO: remove

File: server/src/leihs/procurement/routes.clj [hunk 5]
```diff
@@ -6,11 +6,11 @@
     [clojure.string :refer [starts-with? replace]]
     [clojure.tools.logging :as log]
     [leihs.core.anti-csrf.back :as anti-csrf]
-    [leihs.core.auth.session :as session]
+    [leihs.core.auth.session2 :as session]
     [leihs.core.db :as db]
     [leihs.core.graphql :as core-graphql]
     [leihs.core.http-cache-buster2 :as cache-buster :refer [wrap-resource]]
-    [leihs.core.locale :as locale]
+    [leihs.core.locale2 :as locale]
```
Comment (SeduroDotCom) [inline server/src/leihs/procurement/routes.clj:13]: FYI: needed changes in shared-clj
https://github.com/leihs/leihs-clj-shared/pull/3/files

File: server/src/leihs/procurement/graphql.clj [hunk 6]
```diff
@@ -68,18 +71,18 @@
...
                  (warn "Rolling back transaction because of graphql error: " response)
-                 (jdbc/db-set-rollback-only! tx))
+                 (.rollback (:connectable tx-next)))
                response)
              (catch Throwable th
                (warn "Rolling back transaction because of " th)
-               (jdbc/db-set-rollback-only! tx)
+               (.rollback (:connectable tx-next))
                (throw th))))
       (pure-handler request))))
```
Comment (SeduroDotCom) [inline server/src/leihs/procurement/graphql.clj:87]: FYI: jdbc-transaction-handling

**Comment by SeduroDotCom** [comment]
TODO
--
1. Cleanup deps.edn
2. Remove _tmp-dir
3. Rename my-cast to 'cast-obj-attr'?
   - https://github.com/leihs/leihs-clj-shared/blob/f29828e98a6b6ae206d3c83d9684d66a4f422491/src/leihs/core/utils.clj
4. Issue: jdbc-class-type prefix needed
5. Issue: breaking tests with budget-sort (price-sums), additional sort required

POI's
--
a. Test / Timestamp-handling
   - Added additional check      
   https://github.com/leihs/leihs-procure/pull/125/files#diff-d4c2a23187ad4b5f8b4e3a94e0771853175050ed86fa9a4ae60de725340f1cb9

b. Test / Price-sums (additional sort needed) (5)    
   - https://github.com/leihs/leihs-procure/pull/125/files#diff-660eabeb60ddd0aaae59a4dd431e2d3ab55da7a8875a6bf40a3dbc7488823732

c. JDBC-Next prefix (4)    
   - https://github.com/leihs/leihs-procure/pull/125/files#diff-e2c7d337029ac146a8e76b29247e95f1c9508f6fbf7b9c76f647e340c9770cdf
   - https://github.com/leihs/leihs-procure/pull/125/files#diff-8c70513c2756440d2ea05259be1205972b3103148e8af4000c06785118edf81b 
   - server/src/leihs/procurement/resources/viewers.clj

d. Deleted empty file _> server/src/leihs/procurement/resources/request_shared.clj
e. JDBC-Next transaction-handling
   - server/src/leihs/procurement/graphql.clj
   - https://github.com/leihs/leihs-procure/blob/d0750d12cc2f7e9fde5c98d3189301cb60ac26a6/server/src/leihs/procurement/graphql.clj



Main-PR's
--
- https://github.com/leihs/leihs-procure/pull/125/files#
- https://github.com/leihs/leihs-clj-shared/pull/3/files#

File: server/src/leihs/procurement/graphql.clj (hunk 6 above)
Comment (nimaai) [reply server/src/leihs/procurement/graphql.clj:87]: rebase master. i refactored transaction & and other handling in https://github.com/leihs/leihs/commit/aa80c4fbcc76c503682403e097353517c1af5b00

File: server/src/leihs/procurement/resources/viewers.clj (hunk 1 above)
Comment (nimaai) [reply server/src/leihs/procurement/resources/viewers.clj:28]: in this case the prefix has something to do with the internals of `next.jdbc` and i am afraid there is not much that can be done about it. but generally we are not interested in the `update-count` so one can leave it like it is now on `master`, just do the side-effect (`UPDATE`, `INSERT`) and don't care about the return value.

File: server/src/leihs/procurement/resources/template.clj (hunk 2 above)
Comment (nimaai) [reply server/src/leihs/procurement/resources/template.clj:56]: see https://github.com/leihs/leihs-procure/pull/125#discussion_r1441697603

File: server/src/leihs/procurement/resources/request.clj (hunk 3 above)
Comment (nimaai) [reply server/src/leihs/procurement/resources/request.clj:515]: see https://github.com/leihs/leihs-procure/pull/125#discussion_r1441697603

File: server/src/leihs/procurement/routes.clj (hunk 5 above)
Comment (SeduroDotCom) [reply server/src/leihs/procurement/routes.clj]: FYI: removed all ***2 files again - seems to work now

File: server/src/leihs/procurement/resources/request.clj [hunk 7]
```diff
@@ -288,31 +289,31 @@
...
-  (jdbc/execute! tx
-                 (-> (sql/update :procurement_requests)
-                     (sql/sset data)
-                     (sql/where [:= :procurement_requests.id req-id])
-                     sql/format)))
+  (let [result (jdbc/execute-one! tx (-> (sql/update :procurement_requests)
+                                         (sql/set data)
+                                         (sql/where [:= :procurement_requests.id [:cast req-id :uuid]])
+                                         sql-format))]
+    (list (:next.jdbc/update-count result))))
```
Comment (nimaai) [inline server/src/leihs/procurement/resources/request.clj:308]: do we are about the `update-count` here?

File: server/src/leihs/procurement/utils/sql.clj [hunk 8]
```diff
@@ -49,7 +48,7 @@
 
 (defn select-nest
   [sqlmap tbl nest-key]
-  (helpers/merge-select sqlmap [(types/call :row_to_json tbl) nest-key]))
+  (sql/select sqlmap [[[:row_to_json tbl]] nest-key]))
```
Comment (nimaai) [inline server/src/leihs/procurement/utils/sql.clj:51]: i guess one `[]` wrapping suffice

**Review by nimaai — CHANGES_REQUESTED** [review]
overall looks great !

one should try to get rid of the several `:cast` conversions as they should work through the schema if implemented correctly.

`(:user_id auth-entity)` is a special case as this map is produced outside of graphql schema. but, one could put the cast at the respective place when auth-entity is fetched and then one could also get rid of couple manual `:cast`s. The authenticated entity is put to the request map through a particular ring-middleware. you can take a look...

formatting is not nice at some places, but this can be manually fixed when we format all procure with `cljfmt` later.

File: server/src/leihs/procurement/dashboard.clj [hunk 9]
```diff
@@ -20,77 +23,66 @@
...
-                                 (assoc :cacheKey
-                                          (cache-key dashboard-cache-key bp mc))
-                                 (->> (main-categories/merge-image-path
-                                        tx)))))))]
-               (-> bp
-                   (assoc :main_categories main-cats*)
-                   (assoc :cacheKey (cache-key dashboard-cache-key bp))
-                   (assoc :total_price_cents (sum-total-price
-                                               main-cats*)))))))}))
+     :budget_periods (create-budget bps tx requests dashboard-cache-key main-cats)}))
```
[the remark this answers is not in the document]
Comment (SeduroDotCom) [reply server/src/leihs/procurement/dashboard.clj:88]: Exactly, its easier to read and helpful to run/debug fnc with hardcoded data by logbug

File: server/src/leihs/procurement/resources/request.clj [hunk 10]
```diff
@@ -107,35 +107,35 @@
...
-         (map name)
-         (interleave (vals s-map))
-         (cons :case)
-         (apply sql/call))))
+    (let [map (->> s-map
+                   keys
+                   (map name)
+                   (interleave (vals s-map))
+                   (cons :case))
+          result [[map]]] result)))
```
[the remark this answers is not in the document]
Comment (SeduroDotCom) [reply server/src/leihs/procurement/resources/request.clj:115]: simplified it

File: server/src/leihs/procurement/utils/sql.clj (hunk 8 above)
Comment (SeduroDotCom) [reply server/src/leihs/procurement/utils/sql.clj:51]: this is the minimal setup that works
`(sql/select sqlmap [[:row_to_json tbl] nest-key]))`

File: server/src/leihs/procurement/resources/attachments.clj [hunk 11]
```diff
@@ -35,7 +39,7 @@
           md (-> u-row
                  :metadata
                  to-json
```
[the remark this answers is not in the document]
Comment (SeduroDotCom) [reply server/src/leihs/procurement/resources/attachments.clj:41]: Its needed to **convert PersistentHashMap to String**
```
          p (println ">o> type-metadata0 "   (class (-> u-row :metadata )))

          p (println ">o> type-metadata1 " (-> u-row :metadata to-json))

          p (println ">o> type-metadata2 "   (class (-> u-row :metadata to-json)))

>o> type-metadata0  clojure.lang.PersistentHashMap
>o> type-metadata1  {"PDF:PTEX_Fullbanner":"This is pdfTeX, Version 3.141592-1.20a  (Web2C 7.5.3) kpathsea version 3.5.3","System:File Name":"secd.pdf","System:File Permissions":"-rw-r--r--","PDF:PDF Version":1.4,"File:File Type Extension":"pdf","PDF:Linearized":"No","PDF:Page Count":1,"System:Directory":"/Users/mradl/repos/leihs/procure/server/spec/files","Exif Tool:Exif Tool Version":12.5,"System:File Size":"57 kB","PDF:Creator":"TeX","PDF:Producer":"pdfTeX-1.20a","File:MIME Type":"application/pdf","File:File Type":"PDF","System:File Access Date":"2023-11-09T08:08:00.000+01:00","System:File Inode Change Date":"2023-10-03T10:50:08.000+02:00","System:File Modify Date":"2023-10-03T10:50:08.000+02:00","Source File":"/Users/mradl/repos/leihs/procure/server/spec/files/secd.pdf","PDF:Create Date":"2005-11-28T23:03:50.000-06:00"}
>o> type-metadata2  java.lang.String
```

File: server/src/leihs/procurement/resources/attachments.clj [hunk 12]
```diff
@@ -48,5 +52,5 @@
   [tx ids]
   (jdbc/execute! tx
                  (-> (sql/delete-from :procurement_attachments)
-                     (sql/where [:in :procurement_attachments.id ids])
-                     sql/format)))
+                     (sql/where [:in :procurement_attachments.id (cast-uuids ids)])
```
[the remark this answers is not in the document]
Comment (SeduroDotCom) [reply server/src/leihs/procurement/resources/attachments.clj]: FYI: Removed almost all **:uuid**-casts, and **cast-uuids** isn't in use anymore (removed fnc)

**Pull request merged**